### ISSUE TITLE
[nvmf/vfio-user] Windows 11 fails to initialize NVMe controller

### ISSUE THREAD

**Issue by tmakatos** [spdk/spdk#2576]
# Bug report

Windows 11 fails to initialize the NVMe controller, I see the following in the Device Manager:

```
The device cannot start. (Code 10)

An I/O adapter hardware error has occured.
```

## Context (Environment including OS version, SPDK version, etc.)

Using SPDK 0098e6367.

Windows guest details:
- Edition: Windows 11 Pro
- Version: 21H2
- OS build: 22000.739

**Comment by changpe1** [comment]
@tmakatos I don't have a Windows 11 image, can you post the full log of SPDK?

**Comment by tmakatos** [comment]
[nvmf.log](https://github.com/spdk/spdk/files/9062419/nvmf.log)

**Comment by tmakatos** [comment]
FYI I used QEMU's native NVMe emulation and it worked fine on Windows 11.

**Comment by changpe1** [comment]
@tmakatos Windows 10 1903 still works for me with latest SPDK, and from the nvmf.log I don't see any exceptions. READ command is executed successfully.

Not sure the log is related or not in the quiesce callback:
[2022-07-07 05:48:48.489297] vfio_user.c:2855:vfio_user_log: *DEBUG*: /var/run: closing unexpected fd 449 (index 0) from cmd 8

**Comment by tmakatos** [comment]
> @tmakatos Windows 10 1903 still works for me with latest SPDK, and from the nvmf.log I don't see any exceptions. READ command is executed successfully.

Windows 10 works for me fine as well.

> Not sure the log is related or not in the quiesce callback:
`[2022-07-07 05:48:48.489297] vfio_user.c:2855:vfio_user_log: DEBUG: /var/run: closing unexpected fd 449 (index 0) from cmd 8`

I don't think that's relevant,  I see it in Linux guests and it works fine there.

**Comment by changpe1** [comment]
@tmakatos what's your QEMU command line when starting a Win 11 image?

**Comment by tmakatos** [comment]
> @tmakatos what's your QEMU command line when starting a Win 11 image?

```
qemu-system-x86_64 -cpu host -enable-kvm -smp 4 -m 4G -object memory-backend-file,id=mem0,size=4G,mem-path=/dev/hugepages,share=on,prealloc=yes, -numa node,memdev=mem0 -bios /usr/share/ovmf/OVMF.fd -hda win11_uefi_vdisk.qcow2 -device vfio-user-pci,socket=/var/run/cntrl -machine pc-q35-3.1 -nic user,model=virtio-net-pci -vnc :0
```

**Comment by jimharris** [comment]
Hi @tmakatos - is this issue still valid, or can it be closed?

**Comment by tmakatos** [comment]
@jimharris sorry for the dealyed response, I haven't checked since, has anyone else tried Windows 11?

**Comment by arsiesys** [comment]
I can reproduce this on Windows 10 Pro 22H2.
I was using the master branch for this test and spdk_vhost.

```
    1  rpc.py bdev_nvme_attach_controller -b rdma -t rdma -a xxxx -s 4420 -n nqn.2022-02.io.spdk:cnode1 -f ipv4
    2  rpc.py nvmf_create_subsystem nqn.2019-07.io.spdk:cnode0 -a -s SPDK0
    3  rpc.py nvmf_create_transport -t VFIOUSER
    4  rpc.py nvmf_subsystem_add_ns nqn.2019-07.io.spdk:cnode0 rdman1
    5  rpc.py nvmf_subsystem_add_listener nqn.2019-07.io.spdk:cnode0 -t VFIOUSER -a /vhost/ -s 0

```

I tried the exact same configuration/disk with a Linux VM and it works smoothly.

I am using kubevirt (customized virt-launcher dockage image to use the oracle qemu branch that support vfio-user):
```
  <qemu:commandline>
    <qemu:arg value="-device"/>
    <qemu:arg value="vfio-user-pci,socket=/vhost/cntrl,bus=pcie.0,addr=0xa"/>
  </qemu:commandline>
```

If there is any debug/logs that you may need to help on this case, I will be happy to provide them.

**Comment by tmakatos** [comment]
@arsiesys you can enable debug mode (https://spdk.io/doc/app_overview.html#cmd_arg_log_flags) and attach the log.

**Comment by arsiesys** [comment]
I compiled with the --enabled-debug flag and launched spdk_tgt  with -logflag=all
```
./build/bin/spdk_tgt -R -s 1G -S /tmp/spdk-vhost/ -m0x0000000000000007 --logflag=all 2&>capture.log
```
The log file resulting is 12Go so I filtered it with:
```
cat capture.log | grep -v "context switch" | grep -v "polling all SQs" > filtered.log
```

I started it, execute the commands I did post earlier then, started:
1/ started the VM (it take a while, windows is loading for some minute before starting due to the nvme vfio-user issue).
2/ went in device manager then disabled and re-enabled the nvme storage controller (the one that show the error described in the first post of this thread/issue).

I didn't "catch" anything very helpful in the log, but I post it anyway as you may know better where to look at.
The only weird logs I noticed are:
```
[2024-02-26 20:08:41.683963] vfio_user.c:3089:vfio_user_log: *DEBUG*: /tmp/spdk-vhost/: closing unexpected fd 565 (index 0) from cmd 8
[2024-02-26 20:08:42.109018] vfio_user.c:3089:vfio_user_log: *DEBUG*: /tmp/spdk-vhost/: closing unexpected fd 566 (index 0) from cmd 8
[2024-02-26 20:08:45.838202] vfio_user.c:3089:vfio_user_log: *DEBUG*: /tmp/spdk-vhost/: closing unexpected fd 568 (index 0) from cmd 8
[2024-02-26 20:08:45.873511] vfio_user.c:3089:vfio_user_log: *DEBUG*: /tmp/spdk-vhost/: closing unexpected fd 568 (index 0) from cmd 8
[2024-02-26 20:08:46.210087] vfio_user.c:3089:vfio_user_log: *DEBUG*: /tmp/spdk-vhost/: closing unexpected fd 568 (index 0) from cmd 8
[2024-02-26 20:08:47.950324] vfio_user.c:3089:vfio_user_log: *DEBUG*: /tmp/spdk-vhost/: closing unexpected fd 568 (index 0) from cmd 8
[2024-02-26 20:08:47.985604] vfio_user.c:3089:vfio_user_log: *DEBUG*: /tmp/spdk-vhost/: closing unexpected fd 568 (index 0) from cmd 8
[2024-02-26 20:08:50.158362] vfio_user.c:3089:vfio_user_log: *DEBUG*: /tmp/spdk-vhost/: closing unexpected fd 568 (index 0) from cmd 8
[2024-02-26 20:11:06.884599] vfio_user.c:3089:vfio_user_log: *DEBUG*: /tmp/spdk-vhost/: closing unexpected fd 568 (index 0) from cmd 8
[2024-02-26 20:14:10.792105] vfio_user.c:3089:vfio_user_log: *DEBUG*: /tmp/spdk-vhost/: closing unexpected fd 568 (index 0) from cmd 8
```
[filtered.txt](https://github.com/spdk/spdk/files/14410877/filtered.txt)
Thanks !

**Comment by tmakatos** [comment]
@arsiesys could you:

1. use a bdev_malloc instead of a real NVMe controller (just to keep things simpler)
2. enable logging only for vfio-user (maybe some interesting messages were filtered out)
3. post SPDK logs both for Linux and Windows.

**Comment by tmakatos** [comment]
I compared your log with mine and they're not the same, not sure whether it's because of filtering, there's stuff that's entirely missing in your case, such as "enabling controller".

Also, is Windows fully up to date?

**Comment by arsiesys** [comment]
I collected the logs for a vm with windows and Linux using the following spdk_tgt commands and using malloc instead of an attached nvme device.
```
./build/bin/spdk_tgt -R -s 1G -S /tmp/spdk-vhost/ -m0x0000000000000007 --logflag=nvmf_vfio --logflag=vfio_user 2&>capture.log
```
[linux.txt](https://github.com/spdk/spdk/files/14411848/linux.txt) (Working)
[windows.txt](https://github.com/spdk/spdk/files/14411849/windows.txt) (Not working)


The windows 10 is up to date (I see some KB from February). There is a just a failure to setup the update KB5034441 from january. It seems to be a global issue as I can see on a fast research for that KB.
The windows is a fresh VM (non activated) install that just contains:
- Mellanox OFED drivers (NIC)
- A windows nvmeOF initiator softwre I was planning to experiment

The linux VM is setup with the exact identical specs.

Note that the problem already have impact during windows startup. When the socket device is attached, the boot loading screen of windows take some minute before reaching the login prompt ( like if it was waiting or retrying something then give up).

Indeed. It seems that the line "enabling controller" does appear when using the Linux VM but not the windows one.

**Comment by arsiesys** [comment]
[2024-02-26 21:43:03.653387] => is the moment I stopped the VM/QEMU at the end of the test. That may explain this line at the end of the log file.

I am going to look at what I can get from qemu logs.

**Comment by arsiesys** [comment]
While trying to re-setup the POC without kubevirt to get the logs more easily, I figured out which config create the issue.

```
/usr/libexec/qemu-kvm -smp 4,sockets=1,dies=1,cores=4,threads=1  -vnc :0 -object memory-backend-file,id=mem,size=2G,mem-path=/dev/hugepages,share=on,prealloc=on  -m 2g  -device vfio-user-pci,socket=/tmp/vhost/cntrl,bus=pcie.0,addr=0xa -boot d -cdrom /opt/Win10_22H2_EnglishInternational_x64v1.iso  -machine pc-q35-8.1,usb=off,smm=on,dump-guest-core=off,hpet=off,acpi=on  -drive if=pflash,format=raw,unit=0,file=/usr/share/edk2/ovmf/OVMF_CODE.fd,readonly=on   -drive if=pflash,format=raw,unit=1,file=/usr/share/edk2/ovmf/OVMF_VARS.fd -numa node,memdev=mem,cpus=0-3,nodeid=0
```

Here, when enabling EFI on kubevirt, it will use OVMF from edk2:
```
-drive if=pflash,format=raw,unit=0,file=/usr/share/edk2/ovmf/OVMF_CODE.fd,readonly=on   -drive if=pflash,format=raw,unit=1,file=/usr/share/edk2/ovmf/OVMF_VARS.fd
```
Without this, the window ISO do load and see the vfio-user nvme.

(I honestly forgot at the moment why I had to enable EFI for these POC vm's, but I probably had a reason).

I guess we may still want to understand why vfio-user would not work when windows is using EFI.
The Linux VM does also use EFI (same config) and doesn't have issue with it.

**Comment by tmakatos** [comment]
> Without this, the window ISO do load and see the vfio-user nvme.

@arsiesys does the Windows ISO load _with_ EFI but _without_ vfio-user NVMe?

**Comment by arsiesys** [comment]
Hi!

1) The windows ISO does load with EFI (without waiting time) if I remove the vfio-user device in the qemu arguments.
When vfio-user device is present, it will load after 2m but the disk won't be visible.

2) The linux VM is using the same specs (memory, cpu etc). The disk is available and usable (I was able to run fio on it etc)
I also tried to boot on Ubuntu Server iso and it does load without issue and the disk is visible and even available for the installation.

3) They have the same memory size in the initial POC, but I am able to reproduce locally on my laptop with 2Go memory

4) "bus=pcie.0,addr=0xa" is only necessary when using kubevirt to avoid conflicting with the VGA display pci slot but I reproduce the issue locally without it.

I re-did the exercise by using 1 vcpu. I attached:
1) The windows.txt logs of spdk with annotations inside prefixed by "###" for the timing and the steps on the test.
2) Screen of the device in device manager
[windows.txt](https://github.com/spdk/spdk/files/14437677/windows.txt)
![2024-02-28-112015_720x526_scrot](https://github.com/spdk/spdk/assets/12686720/4889be43-5c77-4540-bc22-dea9742bb39f)


I am also able to setup a reproducible environment of the issue on my laptop with the Windows setup ISO with this workflow:
```
./build/bin/spdk_tgt -R -s 1G -S /tmp/spdk-vhost/ -m0x0000000000000001 --logflag=nvmf_vfio --logflag=vfio_user
./scripts/rpc.py bdev_malloc_create -b malloc0 2500 4096
./scripts/rpc.py nvmf_create_transport -t VFIOUSER
./scripts/rpc.py nvmf_create_subsystem nqn.2019-07.io.spdk:cnode0 -a -s SPDK0
./scripts/rpc.py nvmf_subsystem_add_ns nqn.2019-07.io.spdk:cnode0 malloc0
./scripts/rpc.py nvmf_subsystem_add_listener nqn.2019-07.io.spdk:cnode0 -t VFIOUSER -a /tmp/spdk-vhost/ -s 0
chown 107 /tmp/spdk-vhost/cntrl # Required so the socket is usable by the QEMU in the docker image running on a user uid 107
docker run -v /tmp/spdk-vhost:/tmp/vhost -v /home/XXXXX/Downloads:/opt -v /dev/hugepages:/dev/hugepages --privileged --net=host --entrypoint /bin/bash -it kyon/virt-launcher:dev # kyon/virt-launcher:dev docker image is my forked version of libvirt-launcher using the qemu branch supporting vfio-user
/usr/libexec/qemu-kvm -smp 1,sockets=1,dies=1,cores=1,threads=1  -vnc :0 -object memory-backend-file,id=mem,size=2G,mem-path=/dev/hugepages,share=on,prealloc=on  -m 2g  -device vfio-user-pci,socket=/tmp/vhost/cntrl -boot d -cdrom /opt/Win10_22H2_EnglishInternational_x64v1.iso  -machine pc-q35-8.1,usb=off,smm=on,dump-guest-core=off,hpet=off,acpi=on  -drive if=pflash,format=raw,unit=0,file=/usr/share/edk2/ovmf/OVMF_CODE.fd,readonly=on   -drive if=pflash,format=raw,unit=1,file=/usr/share/edk2/ovmf/OVMF_VARS.fd -numa node,memdev=mem,cpus=0,nodeid=0
```
Note that the same workflow but using the ubuntu server ISO setup does work properly and can see the disk.
I also checked the possibilities of qemu for logging using -d, I tried some options but it does barely log anything :(.

**Comment by ksztyber** [comment]
[Bug scrub] @tmakatos any updates?

**Comment by tmakatos** [comment]
> [Bug scrub] @tmakatos any updates?

I haven't had time to look into it.

**Comment by arsiesys** [comment]
I was curious and retried to run vfio-user with windows setup iso today on my laptop since I noticed than a new qemu fork more up to date was available https://github.com/jlevon/qemu. I also used the master of spdk.

SPDK (using hugepages of 1GB this time)
```
scripts/rpc.py nvmf_create_transport -t VFIOUSER && \
	scripts/rpc.py bdev_malloc_create 20000 512 -b Malloc0 && \
	scripts/rpc.py nvmf_create_subsystem nqn.2019-07.io.spdk:cnode0 -a -s SPDK0 && \
	scripts/rpc.py nvmf_subsystem_add_ns nqn.2019-07.io.spdk:cnode0 Malloc0 && \
	scripts/rpc.py nvmf_subsystem_add_listener nqn.2019-07.io.spdk:cnode0 -t VFIOUSER -a /var/run -s 0
```

QEMU without EFI:
In this case, I will be able to setup windows but it's very slow (maybe because it's only runnig with one CPU).

Note that if I use `-smp 8` to allocate more than one CPU, the windows installer can't create the partitions, in a similar way, If using an already installed windows disk, this one will bsod saying it couldn't mount stuff during boot. So, with windows, this seems to works only with one cpu.
```
sudo ./qemu-system-x86_64 \
-machine accel=kvm,type=q35 -cpu host -m 4G \
-mem-prealloc -object memory-backend-file,id=ram-node0,prealloc=yes,mem-path=/dev/hugepages,share=yes,size=4G \
-numa node,memdev=ram-node0 \
-boot d -cdrom Win10_22H2_EnglishInternational_x64v1.iso \
-device vfio-user-pci,socket=/var/run/cntrl -vnc :0 
```


QEMU with EFI:
Same as before, the installer is very very long until a timeout type of behaviour at boot and do not detect any disk during the setup.
```
sudo ./qemu-system-x86_64 \
-machine accel=kvm,type=q35 -cpu host -m 4G \
-mem-prealloc -object memory-backend-file,id=ram-node0,prealloc=yes,mem-path=/dev/hugepages,share=yes,size=4G \
-numa node,memdev=ram-node0 \
-boot d -cdrom Win10_22H2_EnglishInternational_x64v1.iso \
-device vfio-user-pci,socket=/var/run/cntrl -vnc :0 
-drive if=pflash,format=raw,unit=1,file=OVMF_VARS.fd \
-drive if=pflash,format=raw,unit=0,file=OVMF_CODE.fd,readonly=on
```

**Comment by maxpain** [comment]
Not stale.
Any updates on this?

**Comment by maxpain** [comment]
@tmakatos Please reopen this issue

**Comment by glimchb** [comment]
@maxpain i reopened but it is labeled waiting on submitter . Please provide additional information

**Comment by Kamillaova** [comment]
@glimchb the NVMe controller emulated by SPDK works on linux, with no warnings in dmesg
however, the Windows NVMe driver just can't initialize the device

I followed this guide (but with qemu instead of CH): https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/docs/vfio-user.md

<img width="1022" height="1172" alt="Image" src="https://github.com/user-attachments/assets/82975726-2ea4-49f6-a68f-37006199ff3c" />

<img width="807" height="274" alt="Image" src="https://github.com/user-attachments/assets/d6eca853-5ba8-4cec-82c5-d19f5b50c8d3" />

**Comment by Kamillaova** [comment]
And even on device re-enable I don't see any calls to the `consume_admin_cmd` function in `lib/nvmf/vfio_user.c`
And there is no `disabling controller` -> `enabling controller` messages, like with linux: <img width="962" height="371" alt="Image" src="https://github.com/user-attachments/assets/c2ce679e-0bd7-4deb-9ed4-1ed1e8e4b32b" />

But the device manager freezes for a long time
<img width="447" height="636" alt="Image" src="https://github.com/user-attachments/assets/b5da3d82-7384-47cc-96aa-7daab5c65df2" />

**Comment by maxpain** [comment]
## Root cause found and fix submitted

After extensive debugging with Windows Server 2025, the root cause of `STATUS_DEVICE_CONFIGURATION_ERROR` has been identified:

**`NVMF_VFIO_USER_MSIX_NUM` is set to 512 (derived from `MAX_QPAIRS_PER_CTRLR`).** Windows `stornvme.sys` fails to allocate 512 MSI-X interrupt vectors during PCI resource assignment.

No real NVMe SSD uses 512 MSI-X vectors — typical range is 16–128. The SPDK runtime default queue pair count is 128 (`MAX_QPAIRS/4`).

**Fix:** reduce `NVMF_VFIO_USER_MSIX_NUM` to 128.

Gerrit review: https://review.spdk.io/c/spdk/spdk/+/28328

### What was tested (Cloud Hypervisor + Windows Server 2025):
- ✅ Windows boot from NVMe (vfio-user) as sole disk — `stornvme.sys` Device Manager Status OK
- ✅ Namespace hotplug/unplug on a running controller
- ✅ I/O: diskpart clean, convert GPT, format NTFS, file read/write
- ✅ Linux boot from NVMe (vfio-user)
- ✅ No VMM patches required (tested on unpatched Cloud Hypervisor main and QEMU 10.2)

### Known limitations (not caused by MSIX_NUM, need separate fixes):
- ❌ Hotplug of a **new** NVMe controller at runtime — VMM (CH/QEMU) does not connect to the vfio-user socket on `device_add`. Only controllers present at VM start work.
- ❌ Controller re-plug after hot unplug requires additional SPDK FLR handler fixes (not part of this patch)

**Comment by tomzawadzki** [comment]
Hi @maxpain can you please not edit the comment overriding the past content ? It would be much easier to follow your findings as they progress. 
Besides, please follow the Gerrit flow for submitting updates to changes, rather than abandoning and submitting new patches - https://review.spdk.io/Documentation/intro-gerrit-walkthrough.html#_reworking_the_change

Considering the issue https://github.com/cloud-hypervisor/cloud-hypervisor/issues/7976, where values of 64 and 512 still produced failures. The original patch had the 64 value set - https://review.spdk.io/c/spdk/spdk/+/28311/1/lib/nvmf/vfio_user.c#b82.

How did you arrive at the `128` value for MSI-X vectors ? Is there some resource you could point to that describes this behaviour ?

**Comment by maxpain** [comment]
@tomzawadzki Sorry, I need more time to investigate all remaining issues. Setting `NVMF_VFIO_USER_MSIX_NUM` to 64/128 resolved the boot issues, but didn't fix hotplug.

**Comment by jlevon** [comment]
@maxpain that's fine, we can take changes individually, no need to fix the world all in one!
I'd suggest a separate issue (and separate CRs) for anything over hotplug.

**Comment by maxpain** [comment]
## Correction: root cause is BAR0 size, not MSI-X

My diagnosis in [#issuecomment-4209936252](https://github.com/spdk/spdk/issues/2576#issuecomment-4209936252) was wrong.

### Root cause

`lib/nvmf/vfio_user.c` computes `NVME_REG_BAR0_SIZE = 0x2000` (8 KB). Windows `stornvme.sys NVMeHwFindAdapter` masks BAR0 base with `0xFFFFC000` (enforcing NVMe 1.0e §2.1.10 MLBAR bits 13:04 reserved). With 0x2000 BAR0, PCI can place it at an 8 KB-aligned address that the mask then corrupts. stornvme returns `SP_RETURN_BAD_CONFIG` → `STATUS_DEVICE_CONFIGURATION_ERROR` (Code 10, or STOP 0x7B if boot disk).

Propolis hit the identical bug in 2022 and shipped the identical fix: [oxidecomputer/propolis#126](https://github.com/oxidecomputer/propolis/pull/126) — read that PR for the full mechanism writeup.

### Fix

```diff
-#define NVME_REG_BAR0_SIZE (NVME_DOORBELLS_OFFSET + NVMF_VFIO_USER_DOORBELLS_SIZE)
+#define NVME_REG_BAR0_SIZE 0x4000
```

Doorbells stay at offset `0x1000`; extra 0x2000 is padding. Performance-neutral.

### Test matrix (Windows Server 2025, Cloud Hypervisor `main`, SPDK `v26.01`)

| MSI-X | BAR0 | 1 ctrl boot | Hotplug | 2 ctrls boot |
|:-:|:-:|:-:|:-:|:-:|
| 512 (stock) | 0x2000 (stock) | ❌ | ❌ | ❌ |
| 512 (stock) | **0x4000** | ✅ | ✅ | ❌* |
| reduced | **0x4000** | ✅ | ✅ | ✅ |

`BAR0 = 0x4000` alone fixes every single-controller case with stock MSI-X.

\* With 2 ctrls × 512 MSI-X the failure is hypervisor-side — CH hits `EMFILE` on eventfd clone (default `nofile=1024`), or `SizeLimitExceeded` on `KvmIrqRouting::from_entries` with raised ulimit. Not a SPDK bug.

### MSI-X retraction

@tomzawadzki — your performance pushback on [+/28311](https://review.spdk.io/c/spdk/spdk/+/28311) was correct. Without the BAR0 fix, MSI-X reduction was coincidentally masking the real bug by shortening stornvme's retry loops.

Next I'll reproduce the same matrix under QEMU (instead of Cloud Hypervisor) to confirm the BAR0 fix holds there too.

**Comment by tmakatos** [comment]
Thanks for following up on this, however hardcoding BAR0 to 16 KB is wrong, NVMe supports up to 64K queues so using more than 2K QPs the doorbells won't fit in 16 KB:
```
#define NVMF_VFIO_USER_DOORBELLS_SIZE \
        SPDK_ALIGN_CEIL( \
                (NVMF_VFIO_USER_MAX_QPAIRS_PER_CTRLR * 2 * SPDK_NVME_DOORBELL_REGISTER_SIZE), \
                0x1000)
```

**Comment by tmakatos** [comment]
IIUC we need something like:
```
/*
 * NVMe 2.1.10: base address must be bits 31:14
 */
#define NVME_MLBAR_BA 0x4000
#define NVME_REG_BAR0_SIZE MAX(NVME_MLBAR_BA , (NVME_DOORBELLS_OFFSET + NVMF_VFIO_USER_DOORBELLS_SIZE))
```

**Comment by maxpain** [comment]
Going with your form. For the default `MAX_QPAIRS = 512` it resolves to `0x4000`, which I tested earlier on upstream QEMU 10.2 + SPDK `v26.01` (stock `NVMF_VFIO_USER_MSIX_NUM = 512`), Windows Server 2025 guest: single-controller boot, two controllers at boot, namespace hotplug/unplug, controller hotplug/unplug via QMP — all pass.

One related thing worth flagging: the computed path `NVME_DOORBELLS_OFFSET + NVMF_VFIO_USER_DOORBELLS_SIZE` isn't guaranteed to be a power of two, which PCIe BAR sizing requires. At `MAX_QPAIRS=1024` it yields `0x3000`, at `2048` it's `0x5000`. Your floor masks this at `1024` (lifts to `0x4000`) but not at `≥ 2048`.

Want that addressed in the same CR (wrap in `spdk_align32pow2`) or tracked separately? Happy either way. Once you confirm the shape I'll push the CR.

**Comment by tmakatos** [comment]
> Want that addressed in the same CR (wrap in spdk_align32pow2) or tracked separately? Happy either way. Once you confirm the shape I'll push the CR.

I'd prefer this in a separate PR.

**Comment by maxpain** [comment]
Pushed: https://review.spdk.io/c/spdk/spdk/+/28341 — using your exact form (`spdk_max(NVME_MLBAR_BA, ...)`). The earlier MSI-X CRs (28311, 28328) are now both abandoned with a pointer to this one.
